kong/plugins/ai-proxy/drivers/llama2.lua

```diff
@@ -0,0 +1,291 @@
+            }
+          },
+          object = "text_completion",
+        }
+  
+      end
+
+      return cjson.encode(response_object)
+    end
+  end,
+
+  ["llm/v1/completions"] = function(response_string, route_type)
+    local response_table, err = cjson.decode(response_string)
+    if err then
+      return nil, "failed to decode cohere response"
+    end
+    
+    if (not response_table) or (not response_table.data) or (#response_table.data > 1) then
+      return nil, "cannot parse response from llama2 endpoint"
+    elseif (not response_table.data[1].generated_text) then
+      return nil, "response data is empty from llama2 endpoint"
+    else
+      local split_response = split(response_table.data[1].generated_text, "[/INST]")
+      if not split_response or #split_response < 1 then
+        return nil, "response did not contain a system reply"
+      end
+  
+      local response_object
+
+      -- good
+      if route_type == "llm/v1/chat" then
+        response_object = {
+          choices = {
+            [1] = {
+              message = {
+                content = split_response[#split_response],
+                role = "assistant",
+              },
+              index = 0,
+            }
+          },
+          object = "chat.completion",
+        }
+  
+      elseif route_type == "llm/v1/completions" then
+        response_object = {
+          choices = {
+            [1] = {
+              index = 0,
+              text = split_response[#split_response],
+            }
+          },
+          object = "text_completion",
+        }
+  
+      end
+
+      return cjson.encode(response_object)
+    end
+  end,
+}
+
+function _M.from_format(response_string, model_info, route_type)
+  -- MUST return a string, to set as the response body
+  ngx.log(ngx.DEBUG, "converting from ", model_info.provider, "://", route_type, " type to kong")
+
+  if not transformers_from[route_type] then
+    return nil, fmt("no transformer available from format %s://%s", model_info.provider, route_type)
+  end
+  
+  local ok, response_string, err = pcall(transformers_from[route_type], response_string, route_type)
+  if not ok or err then
+    return nil, fmt("transformation failed from type %s://%s: %s", model_info.provider, route_type, err or "unexpected_error")
+  end
+
+  return response_string, nil
+end
+
+function _M.to_format(request_table, model_info, route_type)
+  ngx.log(ngx.DEBUG, "converting from kong type to ", model_info.provider, "/", route_type)
+
+  if route_type == "preserve" then
+    -- do nothing
+    return request_table, nil, nil
+  end
+
+  if model_info.options.llama2_format == "raw" then
+    -- dynamically call the correct transformer
+    local ok, response_object, content_type, err = pcall(transformers_to[route_type], request_table, model_info.name, model_info.options.max_tokens, model_info.options.temperature, model_info.options.top_p, model_info.options.top_k)
```

Comment:
Nit: we're not _super_ strict about line lengths, but try to keep to 80ish columns when it makes sense. In general our style is to align to opening parens like so:

```lua
    local ok, response_object, content_type, err = pcall(transformers_to[route_type],
                                                         request_table, model_info.name,
                                                         model_info.options.max_tokens,
                                                         model_info.options.temperature,
                                                         model_info.options.top_p,
                                                         model_info.options.top_k)
```

...but in my opinion it's also generally acceptable to use this style:

```lua
    local ok, response_object, content_type, err = pcall(
      transformers_to[route_type],
      request_table,
      model_info.name,
      model_info.options.max_tokens,
      model_info.options.temperature,
      model_info.options.top_p,
      model_info.options.top_k
    )
```

Reply:
I've reformatted to less than 90 char everywhere.